linkerd/http-route/src/http/filter/redirect.rs

```diff
@@ -0,0 +1,416 @@
+use std::num::NonZeroU16;
+
+use super::ModifyPath;
+use crate::http::RouteMatch;
+use http::{
+    uri::{Authority, InvalidUri, PathAndQuery, Scheme, Uri},
+    StatusCode,
+};
+
+#[derive(Clone, Debug, Default, Hash, PartialEq, Eq)]
+pub struct RedirectRequest {
+    pub scheme: Option<Scheme>,
+    pub host: Option<String>,
+    pub port: Option<NonZeroU16>,
+    pub path: Option<ModifyPath>,
+    pub status: Option<StatusCode>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum InvalidRedirect {
+    #[error("redirects may only replace the path prefix when a path prefix match applied")]
+    InvalidReplacePrefix,
+
+    #[error("redirect produced an invalid location: {0}")]
+    InvalidLocation(#[from] http::Error),
+
+    #[error("redirect produced an invalid authority: {0}")]
+    InvalidAuthority(#[from] InvalidUri),
+
+    #[error("no authority to redirect to")]
+    MissingAuthority,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Redirection {
+    pub status: http::StatusCode,
+    pub location: Uri,
+}
+
+// === impl RedirectRequest ===
+
+impl RedirectRequest {
+    pub fn apply(
+        &self,
+        orig_uri: &http::Uri,
+        rm: &RouteMatch,
+    ) -> Result<Option<Redirection>, InvalidRedirect> {
+        let location = {
+            let scheme = self
+                .scheme
+                .as_ref()
+                .or_else(|| orig_uri.scheme())
+                .cloned()
+                .unwrap_or(Scheme::HTTP);
+
+            Uri::builder()
+                .scheme(scheme)
+                .authority(self.authority(orig_uri)?)
+                .path_and_query(self.path_and_query(orig_uri, rm)?)
+                .build()
+                .map_err(InvalidRedirect::InvalidLocation)?
+        };
+        if &location == orig_uri {
+            return Ok(None);
+        }
+
+        let status = self.status.unwrap_or(http::StatusCode::MOVED_PERMANENTLY);
+
+        Ok(Some(Redirection { status, location }))
+    }
+
+    fn authority(&self, orig_uri: &http::Uri) -> Result<Authority, InvalidRedirect> {
+        match (self.host.as_deref(), self.port) {
+            // If a host is configured, use it and whatever port is configured.
+            (Some(h), Some(p)) => format!("{}:{}", h, p).parse().map_err(Into::into),
+            (Some(h), None) => h.parse().map_err(Into::into),
+
+            // If a host is NOT configured, use the request's original host
+            // and either an overridden port or the original port.
+            (None, p) => {
+                let h = orig_uri.host().ok_or(InvalidRedirect::MissingAuthority)?;
+                match p.or_else(|| orig_uri.port_u16().and_then(|p| p.try_into().ok())) {
+                    Some(p) => format!("{}:{}", h, p).parse().map_err(Into::into),
```

Comment:
using the `TryFrom<String>` impl rather than `FromStr` would avoid an allocation/deallocation and is a pretty trivial change here
```suggestion
            (Some(h), Some(p)) => format!("{}:{}", h, p).try_into().map_err(Into::into),
            (Some(h), None) => h.parse().map_err(Into::into),

            // If a host is NOT configured, use the request's original host
            // and either an overridden port or the original port.
            (None, p) => {
                let h = orig_uri.host().ok_or(InvalidRedirect::MissingAuthority)?;
                match p.or_else(|| orig_uri.port_u16().and_then(|p| p.try_into().ok())) {
                    Some(p) => format!("{}:{}", h, p).try_into().map_err(Into::into),
```